plugins/steve/resource-class.ts

```diff
@@ -221,22 +242,44 @@ export function stateDisplay(state) {
   return key.split(/-/).map(ucFirst).join('-');
 }
 
-export function stateSort(color, display) {
+export function stateSort(color: string, display: string) {
   color = color.replace(/^(text|bg)-/, '');
 
   return `${ SORT_ORDER[color] || SORT_ORDER['other'] } ${ display }`;
 }
 
-function maybeFn(val) {
+function maybeFn(val: any) {
   if ( isFunction(val) ) {
     return val(this);
   }
 
   return val;
 }
 
-export default class Resource {
-  constructor(data, ctx, rehydrateNamespace = null, setClone = false) {
+export default class Resource implements ResourceProperties {
+  // Intialize typed properties
+  $ctx = {} as Context;
```

Comment:
Properties of a class that aren't marked as optional ($ctx?: Context) can be initialised in the `constructure`. Things like id could be specifically called out as well `this.id = data.id`

Reply:
I initialized some more values in the constructor. But I believe most of the required props were already being set by this generic code:

```js
constructor(data, ctx, rehydrateNamespace = null, setClone = false) {
    for ( const k in data ) {
      this[k] = data[k];
    }
```

Reply:
Since the context and metadata were already required to be initialized up front, I wondered if the getters for sub-properties of those two things should be used up front.

Then I found this interesting debate over whether getters should ever be used, or if you should always have everything in either a property or method defined on the class. https://github.com/airbnb/javascript/issues/1880 Apparently AirBnB's official guide says getters should never be used.